### ISSUE TITLE
action.d/blocklist.de: tag `<matches>` should get a trailing newline

### ISSUE THREAD

**Issue by faylyn101** [fail2ban/fail2ban#2028]
### Environment:

_Fill out and check (`[x]`) the boxes which apply. If your Fail2Ban version is outdated, 
and you can't verify that the issue persists in the recent release, better seek support 
from the distribution you obtained Fail2Ban from_

- Fail2Ban version (including any possible distribution suffixes):
- OS, including release name/version: 
- [ ] Fail2Ban installed via OS/distribution mechanisms
- [ ] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

This happened in blocklist.de actionban. Read my customization here
```
actionban = curl --fail -s --trace /tmp/curl-<ip>-$$.log -o /dev/null -m 300 --connect-timeout 180 --no-alpn --data-urlencode 'server=<email>' --data 'apikey=<apikey>' --data 'service=<service>' --data 'ip=<ip>'  --data 'format=text' --data-urlencode 'logs=<matches>'$'\n' --user-agent "fail2ban v0.8.12" "https://www.blocklist.de/en/httpreports.html"
```
Look at the last --data-urlencode with "logs", why I added $'\n'? because "https://www.blocklist.de/en/httpreports.html" will sometimes return HTTP 400 Bad Request because it believed additional newline should be appened (no matter it is POST or GET function)

However, this actionban does work in a common shell, but it does not work here in this conf. The captured log from curl says everything.

For example,
```
cat curl-195.208.185.27-23136.log
== Info:   Trying 2a00:1158:2:6d00::2...
== Info: Connected to www.blocklist.de (2a00:1158:2:6d00::2) port 443 (#0)
...
=> Send header, 175 bytes (0xaf)
0000: 50 4f 53 54 20 2f 65 6e 2f 68 74 74 70 72 65 70 POST /en/httprep
0010: 6f 72 74 73 2e 68 74 6d 6c 20 48 54 54 50 2f 31 orts.html HTTP/1
0020: 2e 31 0d 0a 48 6f 73 74 3a 20 77 77 77 2e 62 6c .1..Host: www.bl
0030: 6f 63 6b 6c 69 73 74 2e 64 65 0d 0a 55 73 65 72 ocklist.de..User
0040: 2d 41 67 65 6e 74 3a 20 66 61 69 6c 32 62 61 6e -Agent: fail2ban
0050: 20 76 30 2e 38 2e 31 32 0d 0a 41 63 63 65 70 74  v0.8.12..Accept
0060: 3a 20 2a 2f 2a 0d 0a 43 6f 6e 74 65 6e 74 2d 4c : */*..Content-L
0070: 65 6e 67 74 68 3a 20 33 37 31 0d 0a 43 6f 6e 74 ength: 371..Cont
0080: 65 6e 74 2d 54 79 70 65 3a 20 61 70 70 6c 69 63 ent-Type: applic
0090: 61 74 69 6f 6e 2f 78 2d 77 77 77 2d 66 6f 72 6d ation/x-www-form
00a0: 2d 75 72 6c 65 6e 63 6f 64 65 64 0d 0a 0d 0a    -urlencoded....
=> Send data, 371 bytes (0x173)
0000: 73 65 72 76 65 72 3d 6d 65 25 34 30 6c 69 6f 6e server=me%40lion
...
0050: 37 26 66 6f 72 6d 61 74 3d 74 65 78 74 26 6c 6f 7&format=text&lo
0060: 67 73 3d 4a 61 6e 25 32 30 32 34 25 32 30 30 30 gs=Jan%2024%2000
...
0150: 72 69 6e 67 25 32 30 66 72 6f 6d 25 32 30 31 39 ring%20from%2019
0160: 35 2e 32 30 38 2e 31 38 35 2e 32 37 25 32 34 25 5.208.185.27%24%
0170: 35 43 6e                                        5Cn
== Info: upload completely sent off: 371 out of 371 bytes
== Info: The requested URL returned error: 400 Bad Request
== Info: Closing connection 0
```
Look at the end: "%24%5Cn", which is `$\n`

#### Expected behavior

`$'\n'` should perfectly parsed before sending to actural actionban, not pass it through ...

#### Any additional information

Do you have any best alternative method to correct this? for example, smarter filter to capture exact newline for each line? so that <matches> tag can have enough newline

#### Any customizations done to /etc/fail2ban/ configuration
```
```

#### Relevant parts of /var/log/fail2ban.log file:
_preferably obtained while running fail2ban with `loglevel = 4`_

```
```

#### Relevant lines from monitored log files in question:

```
```

**Comment by faylyn101** [comment]
sorry... HTML escape above actionban example I illustrated

actionban = curl --fail -s --trace /tmp/curl-&lt;ip&gt;-$$.log -o /dev/null -m 300 --connect-timeout 180 --no-alpn --data-urlencode 'server=&lt;email&gt;' --data 'apikey=&lt;apikey&gt;' --data 'service=&lt;service&gt;' --data 'ip=&lt;ip&gt;'  --data 'format=text' --data-urlencode 'logs=&lt;matches&gt;'$'\n' --user-agent &quot;fail2ban v0.8.12&quot; &quot;https://www.blocklist.de/en/httpreports.html&quot;

**Comment by sebres** [comment]
Indeed we can add it and appending of new-line after matches (if blocklist expects this) would be not realy a problem.

- `$'\n'` is a bashism and does not work in posix shell, just try...
```bash
sh -c "echo $'\n'"
```
- as long not fixed, you can try to use something like `$(printf '\n')`.

Additionally, I found another bug here - in current implementation (at least in newest versions, since #1726), where we've fixed potential vulnerability by escaping of the tags, the substitution `'<matches>'` will be interpolated to named parameter `'$f2bV_matches'`, which will not interpolate the this variable, because of single-quotes.
Thus in >= v.0.10 we should rewrite `'<matches>'` to `"<matches>"` for correct substitution of matches.

I'll try to fix it in all our version later... 
Thank you for the note.

**Comment by faylyn101** [comment]
I finally figured out another way to *cheat* blocklist.de and report banned IP without HTTP 400 error. Here is my new actionban in blocklist.de
actionban = curl --fail -s --trace /var/log/fail2ban-curl/curl-&lt;ip&gt;-$$.log -o /dev/null -m 300 --connect-timeout 180 --no-alpn --data-urlencode 'server=&lt;email&gt;' --data 'apikey=&lt;apikey&gt;' --data 'service=&lt;service&gt;' --data 'ip=&lt;ip&gt;'  --data 'format=text' --data-urlencode 'logs=&lt;matches&gt;...
        ...
        ' --user-agent &quot;fail2ban v0.8.12&quot; &quot;https://www.blocklist.de/en/httpreports.html&quot;

The funny thing is actionban can be extended to multiple lines with leading blanks/spaces, which can "simulate" newline symbols. By the way, blocklist.de has also bugs of "logs=" post/get data variable. It at least eats two newlines so that the HTTP 400 would never occur... Weird nginx web server and CGI...

**Comment by sebres** [comment]
Glad it works for you.
I've almost ready a fix for >= 0.10, but I'll not update it for previous versions...
